include/nbl/builtin/hlsl/functional.hlsl

```diff
@@ -64,6 +64,36 @@ ALIAS_STD(less_equal,<=) };
 
 #undef ALIAS_STD
 
+// ------------------------ Compound assignment operators ----------------------
+
+#define COMPOUND_ASSIGN(NAME,OP) template<typename T> struct NAME##_assign { \
+    using type_t = T; \
+    \
+    void operator()(NBL_REF_ARG(T) lhs, NBL_CONST_REF_ARG(T) rhs) \
+    { \
+        lhs = lhs OP rhs; \
```

Comment:
don't take an OP, use the `NAME` functor from above, 

btw see if making the `rhs` template `U` on both `operator()` would break the rest of our shaders

Reply:
It would break the cpp compat, since std::plus, minus etc only template on one parameter

https://github.com/Devsh-Graphics-Programming/Nabla/blob/65d5180d6ad584bd2abb9a087ad21cec18766f2d/include/nbl/builtin/hlsl/functional.hlsl#L17

Reply:
C++ has std::plus<void> with a templated operator(), this would be the closest thing I believe

Reply:
> Best I can do is something like `#define ARITHMETIC_OP(NAME,OP) template<typename LHSType, typename RHSType, typename ReturnType> struct NAME` and have the ones under `ALIAS_STD` be template specializations of this. Same goes for the assign ops
> 
> Reason for this is templating `operator()` seems to not be allowed, and the third template parameter `ReturnType` is needed because we don't have decltype

we do have decltype, instead of `auto` you can use `decltype(declval<RHS>() OP declval<LHS>())` instead of `auto`

Reply:
Do I need to include anything to access the use of decltype?

Reply:
type_traits.hlsl probably, its basically a macro, forwards to `__decltype()` which is the only Clang intrinsic that DXC hasn't killed